backend/cmd/headlamp.go

```diff
@@ -1066,11 +1071,80 @@ func handleClusterHelm(c *HeadlampConfig, router *mux.Router) {
 	})
 }
 
-func handleClusterAPI(c *HeadlampConfig, router *mux.Router) {
+func handleClusterAPI(c *HeadlampConfig, router *mux.Router) { //nolint:funlen,gocognit
 	router.PathPrefix("/clusters/{clusterName}/{api:.*}").HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		var contextKey string
 		clusterName := mux.Vars(r)["clusterName"]
 
-		kContext, err := c.kubeConfigStore.GetContext(clusterName)
+		// checking if kubeConfig exists, if not check if the request headers for kubeConfig information
+		kubeConfig := r.Header.Get("KUBECONFIG")
+		userID := r.Header.Get("X-HEADLAMP-USER-ID")
+		// unique key for the context
+		key := clusterName + userID
+
+		if kubeConfig != "" && c.enableDynamicClusters { //nolint:nestif
+			contexts, err := kubeconfig.LoadContextsFromBase64String(kubeConfig, kubeconfig.DynamicCluster)
+			if err != nil {
+				log.Println("Error setting up contexts from kubeconfig", err)
+				http.Error(w, "Error setting up contexts from kubeconfig", http.StatusBadRequest)
+				return
+			}
+
+			if len(contexts) == 0 {
+				http.Error(w, "Error getting contexts from kubeconfig", http.StatusBadRequest)
+				return
+			}
+
+			for _, context := range contexts {
+				context := context
+
+				if context.Name == clusterName {
+					// check context is present
+					_, err := c.kubeConfigStore.GetContext(key)
+					if err != nil {
+						if err.Error() == "key not found" {
+							if err = c.kubeConfigStore.AddContextWithKeyAndTTL(&context, key, ContextCacheTTL); err != nil {
+								log.Println("Error: failed to store proxy: ", err)
+								http.Error(w, err.Error(), http.StatusInternalServerError)
+								return
+							}
+						}
+					} else {
+						if err = c.kubeConfigStore.UpdateTTL(key, ContextUpdateChacheTTL); err != nil {
+							log.Println("Error: failed to increase ttl: ", err)
+							http.Error(w, err.Error(), http.StatusInternalServerError)
+							return
+						}
+					}
+				}
+			}
+
+			contextKey = key
+		} else {
+			contextKey = clusterName
+		}
+
+		// Parse the URL
+		u, err := url.Parse(r.URL.String())
+		if err != nil {
+			log.Println("Error: parsing URL: ", err)
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return
+		}
+
+		// Get the query parameters
+		queryParams := u.Query()
+
+		// Check if "X-HEADLAMP-USER-ID" parameter is present in the websocket URL.
+		userIDparam := queryParams.Get("X-HEADLAMP-USER-ID")
+		if userIDparam != "" {
+			contextKey = clusterName + userIDparam
+		}
+
+		// Remove the "X-HEADLAMP-USER-ID" parameter from the websocket URL.
+		delete(queryParams, "X-HEADLAMP-USER-ID")
```

Comment:
After the query param is deleted but I dont see the URL being rewritten to request.RawURL after deleting.

Reply:
Updated